Sprint 13

YanikMax opened the pull request.

Commit:
Свёрстан Launch Screen по макету;
Свёрстан главный экран с лентой изображений.

Commit:
Добавлена градиентная подложка под дату.
Экран профиля сверстан с помощью кода.
Добавлен экран отдельного изображения с возможность шеринга картинки.

Commit:
Исправил констрейнты аватара.
Изменил цвет статус-бара на белый.
Полностью убран код связанный с градиентной подложкой.
Для ImageListCell @IBOutlet сделал приватными.
При импорте UIKit убрал Foundation.
Константу ShowSingleImageSegueIdentifier переименовал (не знаю что на меня нашло).
Извелечение опционала в ImagesListViewController реализовал через guard let.
При верстке кодом использовал предложенное флоу.

Commit:
Для входа в приложение пользователь авторизуется через OAuth2.0;
В проекте создан файл Constants.swift, в нём есть константы AccessKey, SecretKey, RedirectURI и AccessScope;
Свёрстан контроллер WebView;
Реализован KVO — приложение показывает, как обновляется прогресс загрузки веб-страницы;
Проверка необходимости авторизации выделена в отдельный контроллер SplashViewController;
Добавлена структура для декодинга JSON-ответа от POST-запроса авторизации Unsplash — OAuthTokenResponseBody;
Создан новый класс OAuth2Service.

Commit:
Переименовал константы, которые были написаны с заглавной буквы;
Исправил в сториборде работу кнопки назад;
В WebViewViewController организовал код согласно общепринятым нормам;
Убрал force unwrapping и заменил на guard
Добавил trailingAnchor, если будет слишком длинное имя;
Добавил константы из файла Constants;
Исправил ключ AccessKey (в начале был лишний символ @)
Приложение было зарегистрировано, но слетели галочки разрешений (исправил).

Commit:
Устранено состояние гонки в OAuthService.
Устранено состояние гонки c помощью блокировки UI.
Сделан сетевой слой для экрана профиля.
Пока идёт загрузка данных профиля, показывается Splash Screen.
Для отображения аватарки пользователя использована библиотека Kingfisher.
В проект добавлен SwiftKeychainWrapper.

Commit:
Добавил использование progressView.setProgress для более плавного заполнения прогресс-бара;
Исправлена опечатка в методе didTapBackButton;
Исправлены названия констант в файле констант;
Перенастроил дата форматтер по макету;
Переименовал константу profile_image в файле ProfileImageService.swift в стиле camelCase;
Переименовал константу DidChangeNotification в файле ProfileImageService.swift в стиле camelCase;
Добавил настройки для аватара, чтобы он был круглым;
Добавил настройки для descriptionLabel, чтобы текст был многострочным и не обрезался;
Вынес значения констант из файла OAuth2Service.swift;
Переписал операцию без force unwrap в файле SplashViewController.swift;
Исправил поток авторизации согласно комментарию;
Переписал операцию без force unwrap в файле ProfileService.swift;
Исправил ссылку для получения профиля в файле ProfileService.swift;
Поля lastName и bio сделал опциональным типом;
Пункт 4.2 в сервис ProfileService не добавлена функция fetchProfile, которая делает запрос GET /me с Bearer Token (но данная функция у меня присутствует в коде).

Commit:
Спасибо Вам, Лилия, за расширенные комментарии.
Сохранение токена перенес в SplashViewController;
Сохранение переписал в SwiftKeychainWrapper;
Изменил путь апи профиля и картинки;
Заменил использование метода на showAuthenticationScreen;
Изменил ключи;
Фон макета изменил на YPBlack;
Констрейнты ширины и высоты убрал;
ProfileService сделано синглтоном.

Commit:
Создан сетевой слой для экрана с лентой фотографий;
Реализована функциональность лайков;
Добавлена возможность полноэкранного просмотра фотографий;
Добавлена возможность выхода из аккаунта.

Commit:
Добавлены [weak self] в ImageListCell в метод setupCell и в SplashViewController в метод fetchProfileImage;
В ImageListService использован тернарный оператор заместо конструкции if-else;
В ImageListService в методы fetchPhotosNextPage и  changeLike добавлены алерты для обработки ошибки;
URL были перенесены в файл Constants;
dateFormatter был перенесен в отдельный extension;
При добавлении наблюдателя Observer так же была добавлена реализация его удаления;
UIBlockingProgressHUD вынесено за пределы switch;
@IBAction func didTapShareButton добавлен private.

Commit:
Проведён рефакторинг вьюконтроллера WebViewViewController;
WebViewController покрыт Unit-тестами;
ProfileViewController и ImageListViewController покрыты Unit-тестами;
Написан UI тест, проверяющий авторизацию в приложение (ДЛЯ РЕВЬЮЕРА!!! Периодически, при тестировании перестает корректно вводить пароль для входа, замедление и простановка паузы после каждого ввода символа не помогает. Наставник помочь не смог, сказав, что иногда бывает такая "магия");
Написан UI тест, проверяющий работу с лентой;
Написан UI тест, проверяющий работу с профилем.

So1ovyov commented on Image Feed/AppDelegate.swift:
```diff
@@ -0,0 +1,42 @@
+@main
+class AppDelegate: UIResponder, UIApplicationDelegate {
+    
+    
```
Лишние пустые строки, можно удалить.

So1ovyov commented on Image Feed/AppDelegate.swift:
```diff
@@ -0,0 +1,42 @@
+    }
+    
+    func application(_ application: UIApplication, didDiscardSceneSessions sceneSessions: Set<UISceneSession>) {
+        // Called when the user discards a scene session.
```
Комментарии можно удалить.

So1ovyov commented on Image Feed/Auth/WebViewPresenter.swift:
```diff
@@ -0,0 +1,53 @@
+import Foundation
+
+//fileprivate let unsplashAuthorizeURLString = "https://unsplash.com/oauth/authorize"
```
Закомментированный код можно удалить.

So1ovyov commented on Image Feed/Auth/WebViewPresenter.swift:
```diff
@@ -0,0 +1,53 @@
+        }
+    
+    func viewDidLoad() {
+//        guard var urlComponents = URLComponents(string: UnsplashAuthorizeURLString) else { return }
```
Закомментированный код можно удалить.

So1ovyov commented on Image Feed/Auth/WebViewPresenter.swift:
```diff
@@ -0,0 +1,53 @@
+        view?.load(request: request)
+        didUpdateProgressValue(0)
+        
+        //webView.load(request)
```
Закомментированный код можно удалить.

So1ovyov commented on Image Feed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,158 @@
+import UIKit
+import WebKit
+
+//fileprivate let unsplashAuthorizeURLString = "https://unsplash.com/oauth/authorize"
```
Закомментированный код можно удалить.

So1ovyov commented on Image Feed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,158 @@
+        webView.navigationDelegate = self
+        presenter?.viewDidLoad()
+        webView.accessibilityIdentifier = "UnsplashWebView"
+        //updateProgress()
```
Закомментированный код можно удалить.

So1ovyov commented on Image Feed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,158 @@
+            forKeyPath: #keyPath(WKWebView.estimatedProgress),
+            options: .new,
+            context: nil)
+        //updateProgress()
```
Закомментированный код можно удалить.

So1ovyov commented on Image Feed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,158 @@
+        }
+    }
+    
+    //    private func updateProgress() {
```
Закомментированный код можно удалить.

So1ovyov commented on Image Feed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,158 @@
+            }
+        }
+    
+    //    private func code(from navigationAction: WKNavigationAction) -> String? {
```
Закомментированный код можно удалить.

So1ovyov commented on Image Feed/AuthConfiguration.swift:
```diff
@@ -0,0 +1,51 @@
+import Foundation
+
+//let unsplashAuthorizeURLString = "https://unsplash.com/oauth/authorize"
```
Закомментированный код можно удалить.

So1ovyov commented on Image Feed/AuthConfiguration.swift:
```diff
@@ -0,0 +1,51 @@
+//    let baseURL = "https://api.unsplash.com"
+//}
+
+let AccessKey = "CrGnzEgtPBgkxTHXzgkRmFYVbe4e7pOKsxyfiUAzMHM"
```
Лучше вынести эти данные в публичный enum и в него добавить константы

So1ovyov commented on Image Feed/AuthConfiguration.swift:
```diff
@@ -0,0 +1,51 @@
+//}
+
+let AccessKey = "CrGnzEgtPBgkxTHXzgkRmFYVbe4e7pOKsxyfiUAzMHM"
+let SecretKey = "JjbQCZd3J1MarTVosM_kTtKph3jJAdAZ5RCol8l4e0k"
```
Нейминг всех констант должен быть с маленькой буквы стиле camelCase

So1ovyov commented on Image Feed/AuthConfiguration.swift:
```diff
@@ -0,0 +1,51 @@
+    let defaultBaseURL: URL
+    let authURLString: String
+    
+    init(accessKey: String, secretKey: String, redirectURI: String, accessScope: String, authURLString: String, defaultBaseURL: URL) {
```
```
    init(
        accessKey: String,
        secretKey: String,
        redirectURI: String,
        accessScope: String,
        authURLString: String,
        defaultBaseURL: URL
    ) {
        self.accessKey = accessKey
        self.secretKey = secretKey
        self.redirectURI = redirectURI
        self.accessScope = accessScope
        self.defaultBaseURL = defaultBaseURL
        self.authURLString = authURLString
    }
```

So1ovyov commented on Image Feed/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,199 @@
+import UIKit
+import Kingfisher
+
+struct UrlsResult: Decodable {
```
Вынести структуру в отдельный файл

So1ovyov commented on Image Feed/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,199 @@
+    let thumb: String
+}
+
+struct PhotoResult: Decodable {
```
Вынести структуру в отдельный файл

So1ovyov commented on Image Feed/Profile/ProfileService.swift:
```diff
@@ -0,0 +1,92 @@
+import Foundation
+
+struct ProfileResult: Codable {
```
Вынести структуру в отдельный файл

So1ovyov commented on Image Feed/Profile/ProfileService.swift:
```diff
@@ -0,0 +1,92 @@
+    }
+}
+
+struct Profile {
```
Вынести структуру в отдельный файл

So1ovyov commented on Image Feed/ServiceAccessLayer/OAuth2Service.swift:
```diff
@@ -0,0 +1,162 @@
+import Foundation
+import WebKit
+
+struct OAuthTokenResponseBody: Decodable {
```
Вынести структуру в отдельный файл

So1ovyov commented on Image Feed/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,199 @@
+            return
+        }
+        let task = urlSession.objectTask(for: request) { (result: Result<PhotoLiked, Error>) in
+            DispatchQueue.main.async {
```
Нужно ослабить self с помощью [weak self]

So1ovyov commented on Image Feed/ImagesList/ImagesListViewController.swift:
```diff
@@ -0,0 +1,193 @@
+    }
+}
+
+extension Collection {
```
Лучше вынести все экстеншены в отдельный файл со структурой Collection

So1ovyov commented on Image Feed/Profile/ProfileImageService.swift:
```diff
@@ -0,0 +1,79 @@
+import Foundation
+
+struct UserResult: Codable {
```
Вынести структуру в отдельный файл

So1ovyov commented on Image Feed/Profile/ProfileImageService.swift:
```diff
@@ -0,0 +1,79 @@
+    }
+}
+
+struct ProfileImage: Codable {
```
Вынести структуру в отдельный файл

So1ovyov commented on Image Feed/SceneDelegate.swift:
```diff
@@ -0,0 +1,44 @@
+    }
+    
+    func sceneDidDisconnect(_ scene: UIScene) {
+        // Called as the scene is being released by the system.
```
Комментарии можно удалить.

Commit:
Убраны лишние пустые строки? задокуменитрованный код и некоторые комментарии;
В AuthConfiguration.swift константы вынесены в публичный enum. Также изменил названия констант в стиле camelCase;
Вынес структуры в отдельную папку Models и покрыл их документацией;
В ImagesListService.swift добавлен [ weak self ] для предотвращения утечки памяти;
Все extensions ImagesListViewController.swift вынесены в отдельный файл вместе со структурой Collections. Были добавлены новые функции в сам класс, так как компилятор не мог определить контекст объектов после вынесения extensions;
В UI-тест были добавлены константы с именем и юзернэймом пользователя и переписана их проверка в тесте c помощью интерполяции строк.

So1ovyov commented on Image FeedUITests/Image_FeedUITests.swift:
```diff
@@ -0,0 +1,76 @@
+import Image_Feed
+
+class Image_FeedUITests: XCTestCase {
+    private let fullName = ""
```
Пустые строки, выдает ошибку в тесте

So1ovyov commented:
Получится сейчас тесты поправить или попробуешь 3-ю попытку на сдачу @YanikMax ?

YanikMax commented:
Да, я сейчас исправлю и отправлю снова. Спасибо Вам.

So1ovyov commented:
Чтобы тест testProfile проходил нужно поменять на следующие данные:
```
private let fullName = "Yandex "
private let userName = "@ypios"
```

Commit:
Поменял данные для констант для ui-теста

So1ovyov commented:
@YanikMax строка Yandex должна быть с пробелом, так в строке располагается никнейм, я дебажил. 
private let fullName = "Yandex "
Можно скопировать прям 2 строки из комментария сверху

Commit:
Внес правку по private let fullName

YanikMax commented:
Максим, прошу прощения. В ревью на практикуме другие были данные. Я их оттуда изначально взял

So1ovyov commented:
Без проблем, возможно как-то в проекте сверстано, так что пробел добавляется а UI тест проверяет и падает. Сейчас все проверю и отпишусь )

YanikMax committed:
Create README.md
